lib/cli/watch-run.js

```diff
@@ -215,23 +216,43 @@ const createWatcher = (
     beforeRun
   });
 
-  watcher.on('ready', async () => {
-    if (!globalFixtureContext) {
-      debug('triggering global setup');
-      globalFixtureContext = await mocha.runGlobalSetup();
-    }
-    rerunner.run();
+  // eslint-disable-next-line no-restricted-globals
+  const startTime = new Date();
+  let ready = false;
+  watcher.on('ready', () => {
+    ready = true;
   });
 
-  watcher.on('all', (event, filePath) => {
-    if (event === 'add') {
-      tracker.regenerate();
+  watcher.on('all', async function handleEvent(event, filePath, stat) {
+    if (exiting) return;
+    if (event === 'unlink' || (stat ? stat.mtime > startTime : ready)) {
+      if (event === 'add') {
+        tracker.regenerate();
+      }
+      if (tracker.has(filePath)) {
+        rerunner.scheduleRun();
+      }
+      return;
     }
-    if (tracker.has(filePath)) {
-      rerunner.scheduleRun();
+    if (!ready && !stat) {
+      fs.stat(filePath, (err, stat) => {
+        if (stat) handleEvent(event, filePath, stat);
+      });
     }
   });
 
+  (async () => {
+    if (!globalFixtureContext) {
```

Comment:
The earlier versions of Mocha using Chokidar 3 still waited until `watcher.on('ready')` to call `mocha.runGlobalSetup()`, but I'm not sure why. A quick look at the code isn't too revealing, but I see why we'd want to call this right away. I'm curious if we can add a test case that has a complex global setup to ensure things still work?

Reply:
I'm not exactly familiar with what that global setup is (loading required modules?) but I can try to figure out what goes into it